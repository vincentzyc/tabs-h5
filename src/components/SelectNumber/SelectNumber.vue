<template>
  <div class="select-number-wrapper">
    <div class="search-number-wrapper">
      <BelongCity :locationCity="locationCity" @selected="handleSelectedBelong" :showCityAll="showCityAll" />
      <SearchNumber @search="handleSearch" />
      <NumberRule @selected="handleSelectedRule" />
    </div>
    <PhoneList :list="phoneList" @handleSelectedNum="lockNumber" />
    <TogglePage
      v-show="phoneList.length > 0"
      :pageIndex="pageIndex"
      :nextLoading="nextLoading"
      @prepage="handlePrepage"
      @nextpage="handleNextpage"
    />
  </div>
</template>

<script setup lang="ts">
import { PageIdLocation, TypeNumItem } from "@/api/types/common";
import "./css/index.styl";
import { useMainStore } from "@/pinia";
import { CommonApi } from "@/api";
import { closeLoading, openLoading } from "@/utils/loading";
import { showToast } from "vant";

const mainStore = useMainStore();

watch(
  () => mainStore.cjData,
  n => {
    if (tabLoad) return;
    if (n) {
      initBelong(n);
      tabLoad = true;
    }
  }
);

const defaultNum = 8;

let allNums: TypeNumItem[] = [];
let selectedBelong = ["", ""];
let searchNum = "";
let numRule = "ALL1";

let tabLoad = $ref(false);
let pageIndex = $ref(1);
let locationCity = $ref([""]);
let showCityAll = $ref("");
let phoneList = $ref<TypeNumItem[]>([]);
let nextLoading = $ref(false);

function getNumParams() {
  if (!mainStore.cjData) return null;
  const numParams = {
    province: selectedBelong[0] || "",
    city: selectedBelong[1] || "",
    pid: mainStore.cjData.pid || mainStore.pid,
    productCode: mainStore.cjData.productCode || "",
    sysOrderId: mainStore.cjData.pageId || "",
    prettyType: numRule,
    searchNum: searchNum,
  };
  return numParams;
}

function initBelong(cjData: PageIdLocation) {
  if (cjData) {
    locationCity = [cjData.province, cjData.city];
    showCityAll = cjData.belongCity;
  }
}

function handleSelectedBelong(belong: string[]) {
  selectedBelong = belong;
  getNumber();
}

function handleSearch(num: string) {
  searchNum = num;
  getNumber();
}

function handleSelectedRule(rule: string) {
  numRule = rule;
  getNumber();
}

function handlePrepage() {
  pageIndex--;
  setPhoneList(pageIndex);
}

function handleNextpage() {
  pageIndex++;
  setPhoneList(pageIndex);
}

function setPhoneList(pageIndex: number) {
  const startIndex = (pageIndex - 1) * defaultNum;
  const endIndex = pageIndex * defaultNum;
  const pagePhoneList = allNums.slice(startIndex, endIndex);
  if (Array.isArray(pagePhoneList) && pagePhoneList.length === 0) {
    nextLoading = true;
    return getNumber();
  }
  phoneList = pagePhoneList.filter(v => !v.isLock);
}

async function lockNumber(numItem: TypeNumItem, key: number) {
  if (!mainStore.cjData) return;
  openLoading("拼命抢号中...");
  const param = {
    pid: mainStore.cjData.pid || mainStore.pid,
    handleNo: numItem.num,
    productCode: mainStore.cjData.productCode || "",
    sysOrderId: mainStore.cjData.pageId || "",
    province: selectedBelong[0] || "",
    city: selectedBelong[1] || "",
  };
  const res = await CommonApi.lockNumber(param);
  closeLoading();
  if (res?.code === "0") {
    handleSelectedNum(numItem);
  } else {
    showToast("您下手太慢了，该号码已被别的用户选取！");
    phoneList.splice(key, 1);
    if (Array.isArray(allNums)) {
      const numsLength = allNums.length;
      for (let index = 0; index < numsLength; index++) {
        const element = allNums[index];
        if (element.num === numItem.num) {
          element.isLock = true;
          break;
        }
      }
    }
  }
}

function handleSelectedNum(numItem: TypeNumItem) {
  console.log(numItem);
  // this.item.value = phone.num;
  //     this.item.selectNum++;
  //     this.item.valueObj = phone;
  //     this.Data.newProductCode = phone.productCode || "";
  //     if (this.item.popupList?.length > 0) {
  //       this.$api.reportMatomo("归属地选号3-点击弹窗");
  //       this.$refs.selectNumCustomPopup.open();
  //     }
  //     this.$api.reportMatomo("选择号码", phone.num);
}

function resetData() {
  pageIndex = 1;
  allNums = [];
  phoneList = [];
  nextLoading = false;
  closeLoading();
}

async function getNumber() {
  pageIndex = 1;
  const param = getNumParams();
  openLoading("拼命加载中...");
  let res = await CommonApi.getPrettyMixItem(param);
  if (res.data && Array.isArray(res.data.numItem) && res.data.numItem.length > 0) {
    allNums = res.data.numItem;
    phoneList = allNums.slice(0, defaultNum);
    nextLoading = false;
    closeLoading();
  } else {
    resetData();
  }
}
</script>
