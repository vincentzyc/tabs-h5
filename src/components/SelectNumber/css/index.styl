$white = #fff;
$black = #000;
$gold = #ffd978;
$gold-d = #fbbf52;
$gold-s = #fade94;
$red = #ff0000;
$gray = #666;
$gray-s = #999;
$gray-ss = #ccc;

.select-number-wrapper {
  .search-number-wrapper {
    padding: 12px;
    border-radius: 10px;
    background-color: $white;
  }

  .select-city {
    width: 100%;
    border: 1px solid $gold;
    border-radius: 8px;
    font-size: 14px;
    padding: 10px;
    margin-bottom: 15px;
  }

  .city-tip {
    margin-right: 24px;
  }

  .city-tip::after {
    content: '';
    position: absolute;
    top: 8px;
    border: 6px solid transparent;
    border-top-color: $gold;
    right: -18px;
  }

  .location-icon {
    -webkit-animation: animte-jump 0.8s linear alternate infinite;
    -ms-animation: animte-jump 0.8s linear alternate infinite;
    -o-animation: animte-jump 0.8s linear alternate infinite;
    animation: animte-jump 0.8s linear alternate infinite;
  }

  .input-search-wrap {
    border: 1px solid $gold;
    border-radius: 10px;
  }

  .number-rule {
    margin-top: 5px;
    border-radius: 8px;
    overflow: hidden;
    color: $white;
    font-size: 14px;
    line-height: 1;

    .rule-item {
      text-align: center;
      width: 24%;
      padding: 12px 0;
      border: 1px solid $gold;
      border-radius: 20px;
      margin-top: 5px;
      transition: 0.3s;
      word-break: break-all;
      color: $black;
      background-color: $gold-d;
      font-size: 14px;
      font-weight: 400;

      &.active {
        color: $white;
      }
    }
  }

  .search-button {
    padding: 0 20px 0 24px;
    background-color: $gold-s;
    background-image: linear-gradient(to right, #f1e5bb, #ffdb85);
    border-radius: 10px;
    line-height: 40px;
    font-size: 18px;
    font-weight: bold;
    color: #473f28;
    transition: 0.3s;
  }

  .search-icon {
    position: absolute;
    top: 10px;
    left: 10px;
    width: 20px;
    color: $gray-ss;
    font-size: 20px;
  }

  .input-search {
    padding-left: 40px;
    -webkit-appearance: none;
    background-color: transparent;
    box-sizing: border-box;
    border: none;
    color: $black;
    font-size: 14px;
    height: 40px;
    line-height: 40px;
    outline: none;

    &::-webkit-input-placeholder {
      color: $gray !important;
    }

    &::-moz-placeholder {
      color: $gray !important;
    }

    &:-moz-placeholder {
      color: $gray !important;
    }

    &:-ms-input-placeholder {
      color: $gray !important;
    }
  }

  .phone-list {
    margin-top: 15px;
  }

  .phone-item {
    min-width: 136px;
    width: 50%;
    position: relative;
  }

  .phone-number {
    position: relative;
    font-size: 22px;
    color: $white;
    border: 1px solid $gold;
    min-width: 136px;
    border-radius: 10px;
    height: 56px;
    line-height: 32px;
    margin: 5px;
    transition: 0.3s;
    background-color: $white;

    &.isactive {
      border-color: $red;
      box-shadow: inset 0 0 10px $red;
    }

    .num-char {
      color: $black;
    }
  }

  .phone-num {
    position: relative;
    border: 1px solid $gold;
    font-size: 18px;
    width: 150px;
    border-radius: 10px;
    line-height: 28px;
    height: 56px;
    background-repeat: no-repeat;
    background-position: 0 0;
    background-size: 100% 100%;
  }

  .phone-info {
    font-size: 12px;
    line-height: 18px;
    width: 100%;
    color: $gray-s;
  }

  .phone-price {
    color: $gray-s;
  }

  .phone-text {
    color: #fd3134;
  }

  .empty-tip {
    font-size: 14px;
    padding: 10px;
    line-height: 24px;
    text-align: center;
    color: $gray-ss;
  }

  .change-button-wrap {
    margin-top: 15px;
  }

  .change-button {
    background-color: $gold-d;
    border-radius: 15px;
    font-size: 18px;
    border: none;
    color: $white;

    .van-icon, .van-button__text {
      font-weight: bold;
    }

    &:not(:last-child) {
      margin-right: 20px;
    }
  }
}