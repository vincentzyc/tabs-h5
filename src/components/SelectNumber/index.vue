<template>
  <div class="select-number-wrapper">
    <div class="search-number-wrapper">
      <BelongCity />
      <SearchNumber />
      <NumberRule />
    </div>
    <PhoneList />
    <TogglePage />
  </div>
</template>

<script setup lang="ts">
import "./css/index.styl";
import { useMainStore } from "@/pinia";
import { watch } from "vue";

const mainStore = useMainStore();

watch(
  () => mainStore.cjAllData,
  (n, o) => {
    console.log(n, o);
  }
);
</script>
