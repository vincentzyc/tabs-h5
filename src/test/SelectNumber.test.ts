import { mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import SelectNumber from '@/components/SelectNumber/SelectNumber.vue'

// function sleep(delay = 0): Promise<void> {
//   return new Promise((resolve) => {
//     setTimeout(resolve, delay);
//   });
// }

test('mount SelectNumber component', async () => {
  expect(SelectNumber).toBeTruthy()

  const wrapper = mount(SelectNumber, {
    global: {
      plugins: [createTestingPinia()],
    },
    props: {

    },
  })

  expect(wrapper.html()).toMatchSnapshot()

  // await wrapper.get('.agreement-confirm').trigger('click')

  // const emitEvent = wrapper.emitted('update:show')

  // expect(emitEvent).toHaveLength(1)
})
