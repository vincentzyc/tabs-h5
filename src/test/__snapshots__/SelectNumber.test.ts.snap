// Vitest Snapshot v1

exports[`mount SelectNumber component 1`] = `
"<div class=\\"select-number-wrapper\\">
  <div class=\\"search-number-wrapper\\">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </div>
  <div class=\\"empty-tip\\"> 哎呦，没有适合您的号码！ <p><small style=\\"color: rgb(204, 204, 204);\\">建议缩短号码字数，例如：88</small></p>
  </div>
  <div class=\\"change-button-wrap\\" style=\\"display: none;\\">
    <div class=\\"flex flex-center\\"><button type=\\"button\\" class=\\"van-button van-button--primary van-button--normal van-button--round van-button--disabled change-button pre-button\\" disabled=\\"\\">
        <div class=\\"van-button__content\\"><i class=\\"van-badge__wrapper van-icon van-icon-arrow-left van-button__icon\\">
            <!---->
            <!---->
            <!---->
          </i><span class=\\"van-button__text\\">上一页</span>
          <!---->
        </div>
      </button><button type=\\"button\\" class=\\"van-button van-button--primary van-button--normal van-button--round change-button next-button\\">
        <div class=\\"van-button__content\\">
          <!----><span class=\\"van-button__text\\">下一页</span><i class=\\"van-badge__wrapper van-icon van-icon-arrow van-button__icon\\">
            <!---->
            <!---->
            <!---->
          </i>
        </div>
      </button></div>
  </div>
</div>"
`;
